File: packages/oxlint-plugin-react-doctor/src/plugin/rules/nextjs/nextjs-no-vercel-og-import.ts
import { defineRule } from "../../utils/define-rule.js";
import type { Rule } from "../../utils/rule.js";
import type { RuleContext } from "../../utils/rule-context.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";

export const nextjsNoVercelOgImport = defineRule<Rule>({
  id: "nextjs-no-vercel-og-import",
  title: "@vercel/og import instead of next/og",
  tags: ["test-noise"],
  requires: ["nextjs"],
  severity: "warn",
  recommendation:
    'Use `import { ImageResponse } from "next/og"`. The `@vercel/og` package is built into Next.js and should not be imported directly',
  create: (context: RuleContext) => ({
    ImportDeclaration(node: EsTreeNodeOfType<"ImportDeclaration">) {
      if (node.source?.value !== "@vercel/og") return;

      context.report({
        node,
        message:
          '@vercel/og is bundled into Next.js. Import from "next/og" instead to avoid duplicate code and version mismatch.',
      });
    },
  }),
});

File: native/oxlint/rules/nextjs-no-vercel-og-import.rs
use oxc_ast::AstKind;
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;

use crate::{AstNode, context::LintContext, rule::Rule};

const MESSAGE: &str = "@vercel/og is bundled into Next.js. Import from \"next/og\" instead to avoid duplicate code and version mismatch.";

#[derive(Debug, Default, Clone)]
pub struct NextjsNoVercelOgImport;

declare_oxc_lint!(
    /// Prefer the Next.js entry point for Vercel OG.
    NextjsNoVercelOgImport,
    react_doctor_native,
    perf,
    version = "0.1.0",
    short_description = "Prefer next/og over @vercel/og.",
);

impl Rule for NextjsNoVercelOgImport {
    fn should_run(&self, ctx: &crate::context::ContextHost) -> bool {
        !is_test_noise_file(ctx)
    }

    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::ImportDeclaration(import_declaration) = node.kind() else {
            return;
        };
        if import_declaration.source.value == "@vercel/og" {
            ctx.diagnostic(OxcDiagnostic::warn(MESSAGE).with_label(import_declaration.span));
        }
    }
}
